state/linklayerdevices_ipaddresses.go

```diff
@@ -227,6 +226,72 @@ func (addr *Address) Remove() (err error) {
 	return addr.st.db().RunTransaction(ops)
 }
 
+// SetOriginOps returns the transaction operations required to set the input
+// origin for the the address.
+// If the address has a provider ID and origin is changing from provider to
+// machine, remove the ID from the address document and the global collection.
+func (addr *Address) SetOriginOps(origin network.Origin) []txn.Op {
+	if addr.Origin() == origin {
+		return nil
+	}
+
+	updates := bson.D{bson.DocElem{Name: "$set", Value: bson.M{"origin": origin}}}
+
+	removeProviderID := origin == network.OriginMachine &&
+		addr.Origin() == network.OriginProvider &&
+		addr.ProviderID() != ""
```

Comment:
What's the purpose of the check on L242? If the provider ID was not populated correctly for some reason but we relinquish control to the machiner shouldn't it assume control of the address regardless of whether the provider ID is set or not?

Reply:
If the origin is changing, we will *always* run at least the update generated on line 238 - setting the origin.

If it happens to be changing back to "machine" and there is a provider ID set, we also update the doc to remove the provider ID *and* we remove the provider ID from the global collection.